quodlibet/qltk/download.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2022 Nick Boultbee
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+
+from os.path import splitext
+from pathlib import Path
+from time import sleep
+from typing import Tuple, Collection, Any, Set
+
+from gi.repository import Soup, GObject
+from urllib3.util import parse_url
+
+from quodlibet import print_d, print_w, _, print_e
+from quodlibet.formats import AudioFile
+from quodlibet.qltk.notif import Task
+from quodlibet.util import http, format_size
+
+
+class DownloadProgress(GObject.Object):
+    """Downloads songs asynchronously, updating a Task"""
+
+    __gsignals__ = {
+        'finished': (GObject.SignalFlags.RUN_LAST, None, (object, object)),
+    }
+
+    def __init__(self, songs: Collection[AudioFile], task=None) -> None:
+        super().__init__()
+        self.songs = songs
+        self.successful: Set[AudioFile] = set()
+        self.failed: Set[AudioFile] = set()
+        self.task = task or Task(_("Browser"), _("Downloading files"))
+
+    def success(self, song: AudioFile) -> None:
+        self.successful.add(song)
+        self._update()
+
+    def failure(self, song: AudioFile) -> None:
+        self.failed.add(song)
+        self._update()
+
+    def _update(self) -> None:
+        frac = self.frac
+        print_d(f"At {frac * 100:.0f}% ({len(self.successful)}, {len(self.failed)})")
+        self.task.update(frac)
+
+    @property
+    def frac(self):
+        frac = (len(self.successful) + len(self.failed)) / len(self.songs)
+        return frac
+
+    def _downloaded(self, msg: Soup.Message, result: Any, data: Tuple) -> None:
+        path, song = data
+        try:
+            headers = msg.get_property('response-headers')
+            size = int(headers.get('content-length'))
+            content_type = headers.get('content-type')
+            print_d(
+                f"Downloaded {format_size(size)} of {content_type}: {song('title')}")
+            _, ext = splitext(parse_url(song("~uri")).path)
+            fn = (song("~artist~title")[:200]
+                  or song("~basename")
+                  or f"download-{hash(song('~filename'))}")
+            path = path / Path(fn + ext)
```

Comment:
Fix: saving files fail if title contains a '/'.

```suggestion
            fn = fn.replace('/', '-')
            path = path / Path(fn + ext)
```

Reply:
Yes thanks. Have escaped all dodgy characters now

Reply:
But not sure it works as expected.

This track from soundcloud `Brain Damage meets Vibronics w/Sir Jean : "Do U remember ? (extended mix)"` is saved to `Vibronics - Brain Damage meets Vibronics w%2FSir Jean %3A %22Do U remember %3F %28extended mix%29%22`. Then this is the title, because title tag is missing.